core/keyboard_nav/key_map.js

```diff
@@ -0,0 +1,250 @@
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview The namespace used to keep track of keyboard actions and the
+ * key codes used to execute those actions.
+ * This is used primarily for keyboard navigation.
+ */
+'use strict';
+
+goog.provide('Blockly.user.keyMap');
+
+/**
+ * Holds the serialized key to key action mapping.
+ * @type {Object<string, Blockly.KeyAction>}
+ */
+Blockly.user.keyMap.map_ = {};
+
+/**
+ * Holds the list of actions that can be executed using the keyboard.
+ * @type {Array<Blockly.KeyAction>}
+ */
+Blockly.user.keyMap.actions_ = [];
+
+/**
+ * Update the key map and actions array to contain the new action.
+ * @param {!string} keyCode The key code serialized by the serializeKeyEvent.
+ * @param {!Blockly.KeyAction} action The action to be executed when the keys
+ *     corresponding to the serialized key code is pressed.
+ * @package
+ */
+Blockly.user.keyMap.setActionForKey = function(keyCode, action) {
+  var key = Blockly.user.keyMap.getKeyByActionName(action.name);
+  // If the action already exists in the key map remove it and add the new mapping.
+  if (key) {
+    delete Blockly.user.keyMap.map_[key];
+  }
+  Blockly.user.keyMap.map_[keyCode] = action;
+  // If the action is not in the actions array add it.
+  if (Blockly.user.keyMap.actions_.indexOf(action) === -1) {
+    Blockly.user.keyMap.actions_.push(action);
+  }
+};
+
+/**
+ * Creates a new key map and a new actions array.
+ * @param {Object<string, Blockly.KeyAction>} keyMap The object holding the key
+ *     to action mapping.
+ * @package
+ */
+Blockly.user.keyMap.setKeyMap = function(keyMap) {
+  var keys = Object.keys(keyMap);
+  // Update the actions array to include any new actions found in the new key map.
+  for (var i = 0; i < keys.length; i++) {
+    var action = keyMap[keys[i]];
+    if (Blockly.user.keyMap.actions_.indexOf(action) === -1) {
+      Blockly.user.keyMap.actions_.push(action);
+    }
+  }
+  Blockly.user.keyMap.map_ = keyMap;
+};
+
+/**
+ * Gets the current key map.
+ * @return {Object<string,Blockly.KeyAction>} The object holding the key to
+ *     action mapping.
+ * @package
+ */
+Blockly.user.keyMap.getKeyMap = function() {
+  return Blockly.user.keyMap.map_;
+};
+
+/**
+ * Get the action by the serialized key code.
+ * @param {string} keyCode The serialized key code.
+ * @return {Blockly.KeyAction|undefined} The action holding the function to
+ *     call when the given keyCode is used or undefined if no action exists.
+ * @package
+ */
+Blockly.user.keyMap.getActionByKeyCode = function(keyCode) {
+  return Blockly.user.keyMap.map_[keyCode];
+};
+
+/**
+ * Get the serialized key that corresponds to the action name.
+ * @param {!string} actionName The name of the action.
+ * @return {string} The serialized key or null if the action does not have
+ *     a key mapping.
+ * @package
+ */
+Blockly.user.keyMap.getKeyByActionName = function(actionName) {
```

Comment:
Why are you using the name of the action instead of the action object itself?

Reply:
Because there are situations in the UI where we only have the action name not the action. Also if you have the action you will always have the action name.